bayes_kit/drghmc.py

```diff
@@ -0,0 +1,471 @@
+            leapfrog_stepcounts
+        )
+        self._damping = self._validate_damping(damping)
+        self._metric = metric_diag or np.ones(self._dim)
+        self._rng = np.random.default_rng(seed)
+        self._theta = (
+            init
+            if (init is not None and init.shape != (0,))
+            else self._rng.normal(size=self._dim)
+        )
+        self._rho = self._rng.normal(size=self._dim)
+        self._prob_retry = prob_retry
+
+        # When generating a draw, the cache is used in the joint_logp() and leapfrog()
+        # functions to avoid recomputing log densities and gradients of the same draw.
+        # Across draws, cache usage depends on the the proposed draw: if accepted,
+        # the cache retains the log density and gradient of the proposed draw; if
+        # rejected, the cache retains the log density and gradient of the current draw.
+        self._cache: list[Tuple[float, VectorType]] = []
+
+    def _validate_propoals(self, proposals: int) -> int:
+        """Validate number of proposals.
+
+        Args:
+            proposals: maximum number of proposal attempts
+
+        Raises:
+            TypeError: proposals is not an int
+            ValueError: proposals is less than one
+
+        Returns:
+            validated number of proposals
+        """
+        if not (type(proposals) is int):
+            raise TypeError(f"proposals must be an int, not {type(proposals)}")
+        if not (proposals >= 1):
+            raise ValueError(
+                f"proposals must be greater than or equal to 1, not {proposals}"
+            )
+        return proposals
+
+    def _validate_leapfrog_stepsizes(
+        self, leapfrog_stepsizes: list[float]
+    ) -> list[float]:
+        """Validate list of leapfrog stepsizes.
+
+        Args:
+            leapfrog_stepsizes: list of non-increasing leapfrog stepsizes
+
+        Raises:
+            TypeError: leapfrog stepsizes is not a list
+            ValueError: leapfrog stepsize list is of incorrect length
+            TypeError: leapfrog stepsize list contains non-float stepsizes
+            ValueError: leapfrog stepsize list contains non-positive stepsizes
+            ValueError: leapfrog stepsize list contains decreasing stepsizes
+
+        Returns:
+            list of validated leapfrog stepsizes
+        """
+        if not (type(leapfrog_stepsizes) is list):
+            raise TypeError(
+                f"leapfrog_stepsizes must be a list, not {type(leapfrog_stepsizes)}"
+            )
+        if len(leapfrog_stepsizes) != self._proposals:
+            raise ValueError(
+                f"leapfrog_stepsizes must be a list of length {self._proposals}, not"
+                f" length {len(leapfrog_stepsizes)}, so that each proposal has a "
+                "specified leapfrog stepsize"
+            )
+        for idx, stepsize in enumerate(leapfrog_stepsizes):
+            if not (type(stepsize) is float):
```

Comment:
Ideally, we want `leapfrog_stepsizes` to accept the following types: `list[int]`, `list[float]`, or `VectorType` (for Numpy arrays). I can change the typing everywhere to enforce this, but it may get messy. Thoughts?